docs/async/async-csharp-vb.rst

```diff
@@ -0,0 +1,200 @@
+	    DisplayData(data);
+	};
+	
+And that's it!  No callbacks and no library to wrestle with.
+
+For those who are more theoretically-inclined, the model used by C# and VB is an implementation of the `Future and Promise concurrency model <https://en.wikipedia.org/wiki/Futures_and_promises>`_.
+
+A few important things to know before continuing:
+
+* Async code relies on ``Task<T>`` and ``Task``, which are constructs used to abstract the notion of a job that you need to wait on for a result.  They also act as the return types, except for the case of Event Handlers (these return ``void``).  `More on Task and Task<T>`_.
+* Marking a method as ``async`` tells the compiler that you're expecting to ``await`` the completion of a task.   Application of ``await`` is optional, but recommended in most cases.
+* ``await`` can only be used inside an async method.
+* When the ``await`` keyword is applied, it suspends the calling method and returns control back to its caller until the awaited task is complete.  This is what makes it responsive.
+* Unless an async method has an ``await`` inside its body, it will never yield!
+* Although not strictly required, ``async void`` should **only** be used on Event Handlers.  More on this later.
+
+Bigger Example (C#)
+------------------
+
+The following example shows somewhat of a "real life" scenario: getting directions to a user for a ridesharing app.
+
+Client app snippet:
+
+.. code-block:: c#
+
+	private async void GetDirectionsForPickup_Pressed(object sender, CustomUserDataEventArgs e)
+	{
+	    var user = e.Source as BasicUserData;
+
+	    // _directionsClient is a made-up object that would encapsulate certain complexity,
+	    // such as getting the location from the device.
+	    var getDirectionsTask = _directionsClient.GetDrivingDirectionsToLocation(user.LastKnownLocation);
+		
+	    // This independent work can be done concurrently
+	    // since it doesn't rely on directions!
+	    NotifyUserOfPickupAsync(user.Contact);
+	    
+	    // The await operator suspends SelectUserForPickup_Pressed, returning control to its caller.
+	    // This is what allows the app to be responsive and not hang on the UI thread.
+	    var directionsJson = await getDirectionsTask;
+		
+	    // Using the JSON.NET library to deserialize data
+	    var directionsData = JsonConvert.DeserializeObject<DirectionsData>(directionsJson);
+	    		    
+	    DrawRouteOnMap(directionsData.Polyline);
+	}
+
+Web service snippet:
+
+.. code-block:: c#
+
+	[HttpGet]
+	public async Task<string> GetDirections(GeoPoint startPos, GeoPoint endPos)
+	{
+	    // Suspends GetUser() to allow the caller (the web service) to accept another request,
+	    // rather than blocking on this one.
+	    var directions = await _directionsService.GetDirectionsAsync(startPos, endPos);
+	    
+	    // Using the JSON.NET library to serialize data
+	    var json = JsonConvert.SerializeObject(directions);
+	    
+	    return json;
+	}
+	
+Bigger Example (VB)
```

Comment:
We should also add F# examples. I see you have a separate doc for F#. We should add a link to it.